### DISCUSSION TITLE
How to use environment variables in route handlers?

### DISCUSSION THREAD

**Discussion by rafaelvascc** [vercel/next.js#69120] [Help] [answered]
### Summary

I'm trying to create an endpoint that returns the app health status and the application version like this:

```
import { NextResponse } from "next/server";

export async function GET(req) {
    return NextResponse.json({ 
        name: "my-app", 
        version: process.env.APP_VERSION || "undefined",
        status: "Healthy",
        checks: []
    });
}
```

It works while I'm debugging on my machine, however when I build and deploy to the server, I get `undefined` in the version property:

![image](https://github.com/user-attachments/assets/8081e1ee-1665-49d8-a5b7-4c1efb36dd21)

So, how can i use environment variables in route handlers? I was digging a little, and looks like nextjs "blocks" access to environment variables in production mode for security reasons and i cant find a way to go around that.
Thanks

**Comment by icyJoseph** [comment, marked as answer]
I can't reproduce. Are these variables available during build time? Or are they made available only for `next start `. Since it is a GET handler, I imagine it is getting statically optimized... and then when the value is made available at runtime, it is "frozen" to undefined.

Try:

```ts
import { NextResponse } from "next/server";

export const dynamic = "force-dynamic";
export async function GET() {
  return NextResponse.json({
    name: "my-app",
    version: process.env.APP_VERSION || "undefined",
    status: "Healthy",
    checks: [],
  });
}
```

That should at least give you the latest available APP_VERSION, when the server is started, without rebuilding.

**Reply by rafaelvascc** [reply]
Not sure what was going on, but `export const dynamic = "force-dynamic"` worked, thanks.
The issue was happening when i was building a container for the application and running on my local machine using docker desktop on windows.
I can share the relevant part of my docker file and docker compose file here:

```
FROM node:20-alpine AS base

COPY ./ /app
WORKDIR /app

FROM base as install
RUN npm install

FROM install as build
RUN npm run build

FROM build as start
CMD npm run start
```

```
services:
  orders-engineering-ui:
    image: app-stuff
    build:
      context: ./
      dockerfile: Dockerfile
    ports:
      - 3000:3000
    container_name: stuff
    environment:
      - APP_VERSION=v0.0.1

**Reply by icyJoseph** [reply]
Yeah, so most likely the env var is not available to the build step. Review your compose and Docker file, this might be affecting other parts of your project. 

And, precisely why dynamic force works, is because at build time, in Next 13 and 14, unless certain conditions are met, GET handlers, are evaluated at build time, and their output, saved to a JSON file, which is served during runtime, rather than evaluating the handler. So called, static optimization.

You can see this one in your build output summary. 

Force dynamic is a way to opt out from that, and there are others. Definitely worth mentioning that, static optimization, will not be the default in Next 15. Somewhat confusingly, the docs seem to be now for 15, which is still RC.

So, the build time doesn't see the env var, and the GET handler is evaluated under those conditions, and the result saved. With force dynamic, the handler is evaluated on demand, at runtime, where the env var is available.
